git/fsck.go

```diff
@@ -0,0 +1,316 @@
+				if !prefixdir.IsDir() {
+					return nil
+				}
+				if len(prefixdir.Name()) != 2 {
+					return nil
+				}
+				objects, err := ioutil.ReadDir(
+					filepath.Join(objdir, prefixdir.Name()),
+				)
+				if err != nil {
+					return err
+				}
+				for _, object := range objects {
+					wantsha1 := fmt.Sprintf("%s%s", prefixdir.Name(), object.Name())
+					oid, err := Sha1FromString(wantsha1)
+					if err != nil {
+						return err
+					}
+					if opts.Verbose {
+						fmt.Fprintf(stderr, "Checking %s %s\n", oid.Type(c), wantsha1)
+					}
+					filename := filepath.Join(objdir, prefixdir.Name(), object.Name())
+					f, err := os.Open(filepath.Join(filename))
+					if err != nil {
+						return err
+					}
+					defer f.Close()
+					zr, err := zlib.NewReader(f)
+					if err != nil {
+						return err
+					}
+					h := sha1.New()
+					if _, err := io.Copy(h, zr); err != nil {
+						return err
+					}
+					sum := h.Sum(nil)
+					sumsha1, err := Sha1FromSlice(sum)
+					if err != nil {
+						// This should never happen, a sha1 from crypto/sha1
+						// should always be convertable to our Sha1 type
+						panic(err)
+					}
+					if sumsha1 != oid {
+						corrupted[oid] = struct{}{}
+						return fmt.Errorf("hash mismatch for %v (expected %v)", filename, oid)
+					}
+				}
+				return nil
+			}()
+			if err != nil {
+				addErr(err)
+			}
+		}
+	}
+
+	var hc []Commitish
+	// Either use RevParse or ShowRef to get a list of all commits that
+	// we want to be checking, depending on if anything was passed as
+	// an argument.
+	if len(objects) != 0 {
+		heads, err := RevParse(c, RevParseOptions{}, objects)
+		if err != nil {
+			addErr(err)
+			// We can't do much more if we can't figure out which objects
+			// we're supposed to be validating.
+			return errs
+		}
+		for _, head := range heads {
+			h, err := head.CommitID(c)
+			if err != nil {
+				addErr(err)
+			}
+			hc = append(hc, h)
+		}
+	} else {
+		heads, err := ShowRef(c, ShowRefOptions{}, nil)
+		if err != nil {
+			addErr(err)
+		}
+		for _, head := range heads {
+			h, err := head.CommitID(c)
+			if err != nil {
+				addErr(fmt.Errorf("not a commit"))
+			}
+			hc = append(hc, h)
+		}
+
+	}
+
+	// Get a list of all reachable objects from the heads.
+	reachables, err := RevList(c, RevListOptions{Quiet: true, Objects: true}, nil, hc, nil)
+	if err != nil {
+		errs = append(errs, err)
+		return errs
+	}
+	for _, obj := range reachables {
+		if opts.Verbose {
+			fmt.Printf("Checking %v\n", obj)
+		}
+		if _, ok := corrupted[obj]; ok {
+			addErr(fmt.Errorf("%v corrupt or missing", obj))
+			continue
+		}
+		o, _, err := c.HaveObject(obj)
+		if err != nil {
+			addErr(err)
+			continue
+		}
+		if !o {
+			addErr(fmt.Errorf("%v corrupt or missing", obj))
+			continue
+		}
+
+		switch ty := obj.Type(c); ty {
+		case "commit":
+			if err := verifyCommit(c, opts, CommitID(obj)); err != nil {
+				addErr(fmt.Errorf("error in commit %v: %v", obj, err))
+			}
+		case "tree":
```

Comment:
Are tags something that could show up in here too?

Reply:
I think so.. there's a few tag related tests which after the commit ones, so it'll probably come up then. First I need to make it past the t1450.15 test, I think, which keeps having a million other little unrelated things that need to be fixed/supported.